pkg/operator/csr/certificate.go

```diff
@@ -0,0 +1,135 @@
+package csr
+
+import (
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"time"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	restclient "k8s.io/client-go/rest"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/klog/v2"
+)
+
+// IsCertificateValid return true if
+// 1) All certs in client certificate are not expired.
+// 2) At least one cert matches the given subject if specified
+func IsCertificateValid(certData []byte, subject *pkix.Name) (bool, error) {
+	certs, err := certutil.ParseCertsPEM(certData)
+	if err != nil {
+		return false, errors.New("unable to parse certificate")
+	}
+
+	if len(certs) == 0 {
+		return false, errors.New("No cert found in certificate")
+	}
+
+	now := time.Now()
+	// make sure no cert in the certificate chain expired
+	for _, cert := range certs {
+		if now.After(cert.NotAfter) {
+			klog.V(4).Infof("Part of the certificate is expired: %v", cert.NotAfter)
+			return false, nil
+		}
+	}
+
+	if subject == nil {
+		return true, nil
+	}
+
+	// check subject of certificates
+	for _, cert := range certs {
+		if cert.Subject.CommonName != subject.CommonName {
+			continue
+		}
+		return true, nil
+	}
+
+	klog.V(4).Infof("Certificate is not issued for subject (cn=%s)", subject.CommonName)
+	return false, nil
+}
+
+// getCertValidityPeriod returns the validity period of the client certificate in the secret
+func getCertValidityPeriod(secret *corev1.Secret) (*time.Time, *time.Time, error) {
+	if secret.Data == nil {
+		return nil, nil, fmt.Errorf("no client certificate found in secret %q", secret.Namespace+"/"+secret.Name)
+	}
+
+	certData, ok := secret.Data[TLSCertFile]
+	if !ok {
+		return nil, nil, fmt.Errorf("no client certificate found in secret %q", secret.Namespace+"/"+secret.Name)
+	}
+
+	certs, err := certutil.ParseCertsPEM(certData)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to parse TLS certificates: %w", err)
+	}
+
+	if len(certs) == 0 {
+		return nil, nil, errors.New("No cert found in certificate")
+	}
+
+	// find out the validity period for all certs in the certificate chain
+	var notBefore, notAfter *time.Time
+	for index, cert := range certs {
+		if index == 0 {
+			notBefore = &cert.NotBefore
+			notAfter = &cert.NotAfter
+			continue
+		}
+
+		if notBefore.Before(cert.NotBefore) {
+			notBefore = &cert.NotBefore
+		}
+
+		if notAfter.After(cert.NotAfter) {
+			notAfter = &cert.NotAfter
+		}
+	}
+
+	return notBefore, notAfter, nil
+}
+
+// BuildKubeconfig builds a kubeconfig based on a rest config template with a cert/key pair
+func BuildKubeconfig(clientConfig *restclient.Config, certPath, keyPath string) clientcmdapi.Config {
```

Comment:
This is currently unused.

Inlining the cert/key might be easier than having to sync all the files to the destination.